ai/AI_working_with_ten.md

```diff
@@ -0,0 +1,898 @@
+```
+
+**Verify:**
+```bash
+curl -s http://localhost:8080/health
+# Expected: {"code":"0","data":null,"msg":"ok"}
+```
+
+### Issue 7: WebSocket Timeout Errors
+
+**Symptoms**:
+```
+Error processing frame: sent 1011 (internal error) keepalive ping timeout
+```
+
+**Cause**: WebSocket connection missing keepalive configuration
+**Solution**: Add ping_interval and ping_timeout
+
+```python
+async with websockets.connect(
+    url,
+    ping_interval=20,
+    ping_timeout=10
+) as ws:
+    # Process frames
+```
+
+### Issue 8: Parallel Audio Routing Crashes
+
+**Symptoms**: Worker crashes when trying to route audio to multiple destinations from intermediate node
+
+**Cause**: TEN Framework doesn't support splitting audio from intermediate nodes
+**Solution**: Split audio at source (agora_rtc)
+
+```json
+{
+  "extension": "agora_rtc",
+  "audio_frame": [
+    {
+      "name": "pcm_frame",
+      "dest": [
+        {"extension": "stt"},
+        {"extension": "my_analyzer"}
+      ]
+    }
+  ]
+}
+```
+
+---
+
+## Quick Troubleshooting Checklist
+
+When something isn't working:
+
+- [ ] Is Docker container running? `docker ps`
+- [ ] Are services running inside container? `curl http://localhost:8080/health`
+- [ ] Check recent logs: `docker exec container_name tail -50 /tmp/task_run.log`
+- [ ] Are environment variables set? Check logs for "Environment variable X is not found"
+- [ ] Did you restart container after editing .env?
+- [ ] Are ports available? Try killing services and restarting
+- [ ] Are Python dependencies installed? `pip3 install ...`
+- [ ] Is extension using correct API? Check for `from ten import` (old) vs `from ten_runtime import` (new)
+- [ ] Are property values being extracted from tuples?
+
+---
+
+## Best Practices
+
+### Security
+- **Never commit API keys** to git
+- Store keys in a persistent file outside the repo (e.g., `/home/ubuntu/api_keys.txt`)
+- Use environment variable placeholders: `${env:VAR_NAME|}`
+- Add `.env` files to `.gitignore`
+
+### Development
+- **Study existing extensions first** - Find similar patterns in existing code before writing new extensions
+- Explore property.json files for connection examples matching your use case
+- Always use `ten_runtime` imports (not `ten`)
+- Extract values from property getter tuples
+- Test health endpoint before testing graphs
+- Check logs frequently during development
+- Use descriptive extension and graph names
+
+### Debugging
+- Start with simple health checks
+- Test individual extensions before complex graphs
+- Use `ten_env.log_info()` liberally for debugging
+- Check session-specific logs for runtime errors
+- Monitor resource usage (memory, CPU) during development
+
+---
+
+## Additional Resources
+
+- [TEN Framework Documentation](https://doc.theten.ai)
+- Python dependencies: `aiohttp`, `pydantic`, `websockets`
+- Agora RTC: Audio/video streaming
+- LLM tool pattern: `AsyncLLMToolBaseExtension`
```

Comment:
`AsyncLLMToolBaseExtension` is not in this repo. It is in ten-framework/ten_ai_base